ssf/samples.go

```diff
@@ -47,13 +69,45 @@ func TimeUnit(resolution time.Duration) SampleOption {
 	}
 }
 
+// Prefix prepends the given prefix to a sample. Due to the
+// by-order-of-arguments nature of applying these arguments, if more
+// than one Prefix argument is given, the leftmost Prefix argument
+// will be the rightmost name component on the final name.
+func Prefix(prefix string) SampleOption {
+	return func(s *SSFSample) {
+		s.Name = prefix + s.Name
+	}
+}
+
 func create(base *SSFSample, opts []SampleOption) *SSFSample {
+	for _, opt := range DefaultSampleOptions {
+		opt(base)
+	}
 	for _, opt := range opts {
 		opt(base)
 	}
 	return base
 }
 
+// Sampled takes a rate and a set of measurements, and returns them as
+// if sampling had been performed: Each measurement gets
+// rejected/included in the result based on a random roll of the RNG
+// according to the rate, and each measurement has its SampleRate
+// field adjusted to match the existing SampleRate * rate.
+func Sampled(rate float32, samples ...*SSFSample) []*SSFSample {
+	res := make([]*SSFSample, 0, len(samples))
+
+	for _, s := range samples {
+		if rand.Float32() <= rate {
+			if rate > 0 && rate <= 1 {
+				s.SampleRate = s.SampleRate * rate
```

Comment:
I'm not sure I understand this part - why do we want to update the internal sample rate with each metric?

Reply:
That's the sample rate reported to the upstream server - it needs to know by what number to multiply the measurement-as-reported in order to have an accurate estimate of the real number. Flow is as follows:

1. client calls `Sampled` on a measurement (a counter with `Value=30` `SampleRate=1`), `rate=0.5`
2. There's a 50% chance that each measurement gets dropped
3. The measurements that don't get dropped have their `SampleRate` adjusted to represent the chance that they got reported; in this example, that's `1 * 0.5`
4. The not-dropped measurements get reported to veneur, which does the inverse computation, [`Value * (1/SampleRate)`](https://sourcegraph.com/github.com/stripe/veneur@4b90543807ff43e1e8ccaecbfa4c168726d82857/-/blob/samplers/samplers.go#L133) to arrive at an estimate of a realistic counter value, taking into account the dropped measurements.

Summing up, I think this way of computing the Sample's SampleRate should allow for things like single and multiple calls of `Sampled` - e.g. `Sampled(0.2, Sampled(0.5, []*ssf.*SSFSample{ssf.Count("foo", 20, nil)}))`

Reply:
Thanks for asking this, @aditya-stripe. I think I also wanted to but got distracted. The explanation makes sense. In effect this is the same as the "can I use different sample rates when calling `Increment`" that we get sometimes. Sure, you can since each metric is multiplied by it's sample rate if it "wins" and makes it to the Veneur's internals. Is that a correct explanation?

Reply:
Exactly! With `Sampled`, you pre-filter the measurements, then just report whatever comes out. I haven't gotten a whole lot of examples of this in (we report at sample rate = 1 almost everywhere except the places where it could lead to amplification), but something like this is totally plausible:

``` go
span.Samples = append(span.Samples, ssf.Sampled(0.9, ssf.Count("rare.counter", 1, nil)...)...)
span.Samples = append(span.Samples, ssf.Sampled(0.01, ssf.Count("expenso.counter", 1, nil)...)...)
```